### PR TITLE
Escape invalid characters when used in Java identifiers

### PR DESCRIPTION
**Pull request by kelnos**
I ran into this when I wanted an enum value called `integer?`, and figured I'd add code for escaping all the various non-identifier characters, at least the ones found on a US keyboard (note that `$` is actually legal in Java identifiers, so it's not included here).

This turned out to be a huge rabbit hole, because, despite the work we did last year to make naming better, there was still a lot of inconsistency and hackery around naming things.  In order to make this a reasonable problem to solve, I added (to `LanguageTerms`) some new methods:

* `formatPackageName`
* `formatTypeName`
* `formatMethodName`
* `formatMethodArgName`
* (we already had) `formatEnumName`

The bonus here is that as we add other languages with different naming conventions, it becomes trivial to have guardrail follow those naming conventions.

In addition, I set things up so some names (for example, the name of a response class) are generated by the core (by calling the above methods) and threading them into the interpreter methods so the interpreters don't keep rebuilding names in an error-prone way.

This doesn't fix everything.  There are still some instances of `.capitalize` and `.uncapitalized` here and there.  There are *fewer* instances than before, which is nice.  And the core no longer does `.toCamelCase`, `.toSnakeCase`, or `.toPascalCase` anywhere.  The one thing I am still not able to eradicate is `ProtocolGenerator`'s `transformProperty`'s `needsCamelSnakeConversion`, because doing so still breaks the conflicting-name fixer-upper, and making that continue to work would require more surgery.  So, sadly, I will yet again punt on that.

This will unfortunately require some user code to change to continue to compile, which is detailed in MIGRATING.

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

**Comment by kelnos** [comment]
Oh, also interesting, perhaps: `BacktickTest` is nearly obsolete, as almost all the names generated no longer need backticks.  Strangely `dashy-class` is still there as-is, which I should perhaps dig into and fix before we merge this.

**Comment by kelnos** [comment]
Ok, two things:

1.  Fixed up protocol definition class names not getting formatted.  I'm not sure I did it in the best way, so I left it in its own commit so you can take a look at it separately.
2.  FINALLY I got rid of `needCamelSnakeConversion`.  Super happy this is gone, finally finally finally.

**Comment by kelnos** [comment]
Oh -- now `BacktickTest` doesn't actually have any backticks in the generated code anymore, heh.  Should we remove it?

Also, apologies for the size of this PR... but I guess most of it is test changes/fixes.

File: modules/codegen/src/test/scala/core/issues/Issue126.scala [hunk 1]
```diff
@@ -31,30 +31,30 @@ class Issue126 extends AnyFunSuite with Matchers with SwaggerSpecRunner {
 
     val handler  = q"""
       trait StoreHandler {
-        def getRoot(respond: StoreResource.getRootResponse.type)(): scala.concurrent.Future[StoreResource.getRootResponse]
+        def getRoot(respond: StoreResource.GetRootResponse.type)(): scala.concurrent.Future[StoreResource.GetRootResponse]
```
Comment (blast-hardcheese) [inline modules/codegen/src/test/scala/core/issues/Issue126.scala:34]: _ah._ This is what you meant. Yes, this is quite unfortunate.

File: modules/sample-dropwizard/src/test/scala/core/Dropwizard/JavaInvalidCharacterEscapingTest.scala [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package core.Dropwizard
+
+import org.scalatest.freespec.AnyFreeSpec
+import org.scalatest.matchers.must.Matchers
+
+class JavaInvalidCharacterEscapingTest extends AnyFreeSpec with Matchers {
+  "Invalid characters should be escaped to reasonable things" in {
+
```
Comment (blast-hardcheese) [inline modules/sample-dropwizard/src/test/scala/core/Dropwizard/JavaInvalidCharacterEscapingTest.scala:8]: the best kind of tests are the ones that can't be written, well done.

File: modules/codegen/src/main/scala/com/twilio/guardrail/ClientGenerator.scala [hunk 3]
```diff
@@ -34,28 +34,26 @@ object ClientGenerator {
...
-          val routes     = unsortedRoutes.sortBy(r => (r.path.unwrapTracker, r.method))
-          val clientName = s"${className.lastOption.getOrElse("").capitalize}Client"
-          def splitOperationParts(operationId: String): (List[String], String) = {
-            val parts = operationId.split('.')
-            (parts.drop(1).toList, parts.last)
-          }
-
+          val routes = unsortedRoutes.sortBy(r => (r.path.unwrapTracker, r.method))
           for {
+            clientName <- formatTypeName(className.lastOption.getOrElse(""), Some("Client"))
```
Comment (blast-hardcheese) [inline modules/codegen/src/main/scala/com/twilio/guardrail/ClientGenerator.scala:46]: I think this is a great opportunity to push the string `"Client"` into each generator, further reducing the impedance mismatch when moving to support non-pascal-case languages.

File: modules/sample-dropwizard/src/test/scala/core/Dropwizard/JavaInvalidCharacterEscapingTest.scala (hunk 2 above)
Comment (kelnos) [reply modules/sample-dropwizard/src/test/scala/core/Dropwizard/JavaInvalidCharacterEscapingTest.scala:8]: Lol oops.  Too much multitasking.

File: modules/codegen/src/main/scala/com/twilio/guardrail/ClientGenerator.scala (hunk 3 above)
Comment (kelnos) [reply modules/codegen/src/main/scala/com/twilio/guardrail/ClientGenerator.scala:46]: This is fine, actually, and i even could have made the param value "client" or "CLIENT" or whatever; the formatter is responsible for doing the right thing for the language.  I wanted the generators to _not_ be responsible for naming things, as much as is reasonably possible.

**Comment by kelnos** [comment]
I'll try to take a stab at the scalafix rules, but might not get to it before you have time.

File: modules/codegen/src/main/scala/com/twilio/guardrail/ClientGenerator.scala (hunk 3 above)
Comment (blast-hardcheese) [reply modules/codegen/src/main/scala/com/twilio/guardrail/ClientGenerator.scala:46]: makes sense, though hopefully my reticence towards hard-coded strings in core is understandable

**Comment by kelnos** [comment]
I'm sure this will need a couple iterations to get right, but here is the scalafix rule I whipped up: https://github.com/blast-hardcheese/guardrail-scalafix-rules/pull/1

**Pull request merged**